### PR TITLE
Benchmark updates

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?

Adds some updates for nd4j benchmarking to add memory pressure test and updated documentation for using the benchmark with jemalloc

Adds opaquendarray caching and deletion on close()
Add new reduce smoke test
## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

### CODE REVIEW

**Review by treo — DISMISSED** [review]

File: contrib/benchmarking_nd4j/pom.xml [hunk 1]
```diff
@@ -83,8 +88,8 @@
                 <version>3.8.1</version>
                 <configuration>
                     <compilerVersion>${javac.target}</compilerVersion>
-                    <source>${javac.target}</source>
-                    <target>${javac.target}</target>
```
Comment (treo) [inline contrib/benchmarking_nd4j/pom.xml:90]: This looks a bit off, you bothch change the parameter as part of the properties, and then change it here to a different fixed value. I guess that isn't quite intentional?

File: contrib/benchmarking_nd4j/src/main/java/org/nd4j/memorypressure/MemoryPressureTest.java [hunk 2]
```diff
@@ -10,12 +10,15 @@ public class MemoryPressureTest {
...
+            INDArray arrOne = Nd4j.ones(200);
+            INDArray arrTwo = Nd4j.ones(200);
             threads[i] = new Thread(() -> {
                 AtomicInteger atomicInteger = new AtomicInteger(0);
                 while (atomicInteger.incrementAndGet() < 100000) {
-                    INDArray arrOne = Nd4j.ones(200);
-                    INDArray arrTwo = Nd4j.ones(200);
+
                     arrOne.addi(arrTwo);
+                    System.out.println("Completed " + atomicInteger.get() + " iterations");
```
Comment (treo) [inline contrib/benchmarking_nd4j/src/main/java/org/nd4j/memorypressure/MemoryPressureTest.java:20]: Is this intentional that you create 100000 prints here?

File: platform-tests/bin/java [hunk 3]
```diff
@@ -23,6 +23,7 @@
 set -exo pipefail
 #JAVA_CALL="java -DsourceCodeIndexerPath=/home/agibsonccc/Documents/GitHub/deeplearning4j -javaagent:/home/agibsonccc/.m2/repository/org/nd4j/nd4j-log-analyzer/1.0-SNAPSHOT/nd4j-log-analyzer-1.0-SNAPSHOT.jar "
 JAVA_CALL="java "
+source /home/agibsonccc/Documents/GitHub/deeplearning4j/platform-tests/bin/env.sh
```
Comment (treo) [inline platform-tests/bin/java]: It seems wrong to keep your path here.

**Review by treo — CHANGES_REQUESTED** [review]

File: contrib/benchmarking_nd4j/src/main/java/org/nd4j/memorypressure/MemoryPressureTest.java (hunk 2 above)
Comment (agibsonccc) [reply contrib/benchmarking_nd4j/src/main/java/org/nd4j/memorypressure/MemoryPressureTest.java:20]: Yes this one runs for a LONG time and the main use case isn't throughput here but GC.

File: platform-tests/bin/java (hunk 3 above)
Comment (agibsonccc) [reply platform-tests/bin/java]: Ah yeah this was left over from some troubleshooting. I wasn't sure if the variables were working or not so I was double checking paths..

**Review by treo — APPROVED** [review]

**Pull request merged**